[Sol] Step3-04 사진 추가하기 기능 구현

Hansolkkim opened the pull request:
## 실행 화면
![SS 2022-03-14 PM 10 29 23](https://user-images.githubusercontent.com/92504186/158181980-bdb713b5-62e8-4fee-9475-b2eff491ce67.gif)
---
## 작업 목록
- [x] Refactoring
  - [x] namespace 활용하여 Plane 객체 내의 Constasnts 정리
  - [x] 팩토리 메서드의 반환 타입을 추상 타입으로 수정
  - [x] BackgroundColor의 입력값 범위를 제한하여 컴파일 단계에서 해당 범위가 맞는지 확인 가능하도록 수정
- [x] UIImagePicker를 이용해 사진을 추가하는 기능 구현
- [x] Photo, PhotoView 클래스 구현
- [x] VC에서 Photo 객체에 대한 입출력 흐름 추가
- [x] Photo를 터치하면 배경색 버튼이 비활성화되는 기능 추가
---
## 고민과 해결
1. 드디어 BackgroundColor 타입의 속성 값들의 제한을 설정할 방법을 생각해냈습니다. UInt8 타입을 이용하면, 0~255 범위 이외의 값들은 컴파일 단계에서 걸러진다는 것을 생각해내 BackgroundColor의 속성인 r,g,b를 모두 UInt8 타입으로 설정하고, BackgroundColor를 초기화 생성자의 매개변수도 모두 UInt8 타입으로 받도록 했습니다.
2. NotificationCenter를 통해 받아온 userInfo를 프로토콜인 추상 타입으로 타입 캐스팅을 하게 되면 Dictionary의 Key 값으로 이용할 수 없게 되어, 다른 방법을 찾다가 Photo를 Rectangle의 자식 클래스로 만들고, 어쩔 수 없이 userInfo를 구체 타입인 Rectangle로 타입 캐스팅하여 문제를 해결할 수 있었습니다. 해결된 것은 아니고 임시 방편을 찾은것 같다는 느낌을 받아, 해당 부분을 계속 고민해가며 수정할 예정입니다.
---
## 질문사항
- 아래의 부분에서 JK께서 프로토콜을 리턴받아놓고 as? RectangleView로 타입 캐스팅하는게 불편해보인다고 방법을 여쭤보셨었는데, 타입 캐스팅 과정 없이는 프로토콜을 리턴 받은 값으로 addSubview(_:)가 불가능했습니다. 해당 부분을 해결하기 위해, RectangleView를 생성해주는 팩토리 메서드의 반환 타입을 `UIView & RectangleViewable` 로 수정했습니다. 혹시 이런 경우에도 팩토리 메서드의 반환 타입이 구체 타입을 반환한다고 볼 수 있는건가요?
  - 수정전
  ```swift
  // class ViewFactory
  static func makeRectangleView(of rectangle: Rectangle) -> RectangleView {...}
  // class ViewController
  // @objc func planeDidAddRectangle(_:) {
   guard let newRectangleView = ViewFactory.makeRectangleView(of: addedRectangle) as? RectangleView else {return}
   self.drawableAreaView.addSubview(newRectangleView)
  ```
  - 수정후
  ```swift
  // class ViewFactory
  static func makeRectangleView(of rectangle: Rectangle) -> UIView & RectangleViewable {...}
  // class ViewController
  // @objc func planeDidAddRectangle(_:) {
  let newRectangleView = ViewFactory.makeRectangleView(of: addedRectangle)
  self.drawableAreaView.addSubview(newRectangleView)
  ```
- notificationCenter로 부터 userInfo를 받아오는 과정에서, 받아온 userInfo의 필요한 키의 Value를 `as? Rectangle` 로 변환해주다보니, 구체 타입에 의존한다는 리뷰를 받았는데, 이번 스텝을 진행하다보니 이 부분을 추상 타입으로 변환해주어야 Photo, Label 등을 일괄적으로 한 번에 처리할 수 있겠다는 생각을 했습니다. 
  이 부분에서 저는, 제가 선언한 Rectangularable 프로토콜을 채택하는 모델(예를 들어 RectModel로 이름을 설정하겠습니다.)을 선언하고, Photo, Label, Rectangle 객체가 RectModel을 상속받는 경우에는 이 문제를 해결할 수 있을 것 같은데,  `as? RectModel` 로 변환을 한다고 해서 이 부분에서 구체 타입에 의존하지 않지는 않기 때문에 말씀해주신 문제를 해결하는 방법은 아닌 것 같습니다. 그렇다고 프로토콜로 변환을 해주다보니, matchedView를 찾는 과정에서 [AnyHashable: RectangleViewable] 딕셔너리에 Key로 notificationCenter로부터 userInfo로 받아온 값을 사용해야 하는데, 추상 타입은 Hashable 할 수 없기 때문에 matchedView를 찾는 곳에서 문제가 생겼습니다.
  이런 경우에, 제가 말씀드린 RectView를 이용해 userInfo의 특정 Value를 변환하는 방법이 추상 타입에 의존한다고 볼 수 있을까요?
  리뷰 달아주셨던 코드는 아래의 코드입니다.
  ```swift
  @objc func planeDidSpecifyRectangle(_ notification: Notification) {
    guard let specifiedRectangle = notification.userInfo?[Plane.UserInfoKeys.specifiedRectangle] as? Rectangle,
          let matchedView = rectangleAndViewMap[specifiedRectangle] else {
              initializeViewsInTouchedEmptySpaceCondition()
              return
          }
  ...
  }
  ```

Hansolkkim committed:
Refactor: namespace 활용해 Plane 내의 Constants 정리

Hansolkkim committed:
Refactor: Plane의 subscript 수정

subscript의 의미에 어울리도록 다시 구현

Hansolkkim committed:
Refactor: ViewController, Plane의 notificationCenter 변수 삭제

문제를 야기할 수 있는 싱글톤에 대한 참조 변수 제거

Hansolkkim committed:
Refactor: ViewController에서 사용할 Float 타입 변환을 Float 타입를 확장해 normalized()  메서드로 구현

Float을 계산하는 과정에서 불필요한 값들이 추가되는 과정이 있어, 해당 값의 불필요한 부분을 버리도록하는 normalized() 메서드 구현

Hansolkkim committed:
Refactor: ViewFactory 메서드의 반환 타입 수정

drawableAreaView에 서브뷰로 ViewFactory 메서드의 반환값을 이용하기 위해, ViewFactory 메서드의 반환 타입을 UIView & RectangleViewable로 수정

Hansolkkim committed:
Refactor: updateSeletedView(_:) 메서드의 매개변수가 추상 타입에 의존하도록 변경

해당 매개변수가 추상 타입에 의존할 수 있도록, RectangleViewable 프로토콜의 변수를 추가적으로 선언

Hansolkkim committed:
Refactor: BackgroundColor 타입 초기화시 입력 제한 설정을 위한 내부 프로퍼티의 타입 변경

내부 프로퍼티 r, g, b의 타입을 UInt8로 설정해, 0~255 범위 내의 정수로만 BackgroundColor 타입이 초기화되도록 설정

Hansolkkim committed:
Refactor: RectangleViewable 프로토콜 수정으로 인한 타입 캐스팅 부분 제거

RactangleViewable 프로토콜이 프로퍼티들을 갖게 됨으로 인해 matchedView 상수를 타입 캐스팅할 필요가 없어져, `as? RectangleView` 의 타입 캐스팅 부분 제거

Hansolkkim committed:
Docs: README.md 파일 업데이트

Hansolkkim committed:
Feat: 사진을 추가할 수 있는 새로운 버튼 추가

새로운 버튼으로 인해 사각형 버튼의 위치 변경
버튼들의 기능을 설명하기 위해 버튼이 SFSymbol을 이용해 모양을 보여주도록 설정
버튼들을 좀 더 둥글게 표현하도록 button.layer.cornerRadius 값 설정
사진을 추가하는 버튼의 UIAction은 ImagePicker를 띄워주도록 설정

Hansolkkim committed:
Feat: 사진을 포함하는 사각형 객체를 구현하기 위해 Photo 클래스 선언

Plane, ViewController에서 Photo와 Rectangle을 한꺼번에 처리가 가능하도록 Photo 클래스가 Rectangularable 프로토콜을 채택
RectangleFactory에 새로운 Photo 인스턴스를 생성하는 makePhoto(in:image:) 메서드 선언

Hansolkkim committed:
Refactor: Photo의 image 속성을 Data 타입으로 변경

Photo 객체에서는 이미지에 대한 정보를 data로 갖고 있도록 설정

Hansolkkim committed:
Feat: Plane 객체에서 새로운 Photo 객체를 생성해 rectangles 속성에 추가하는 addNewPhoto(in:with:) 메서드 선언

해당 Photo 객체가 생성되서 추가되었음을 NotificationCenter에 알리기 위한 NotificationNames, UserInfoKeys 값 추가

Hansolkkim committed:
Feat: Photo에 대응하는 PhotoView 클래스 선언

PhotoView 객체도 RectangleView 객체와 함께 처리될 수 있도록 RectangleViewable 프로토콜을 채택
ViewFactory 클래스 내에 PhotoView를 생성해내는 makePhotoView(of:) 메서드 추가

Hansolkkim committed:
Feat: ViewController에 Photo 객체에 대한 입출력 흐름 추가

Plane에서의 Photo 객체 추가에 대한 Observer 추가
Plane에서 Photo 객체 추가에 대한 노티가 오면 이 이벤트를 처리하기 위한 planeDidAddPhoto(_:) 메서드 추가
ViewController에서 ImagePicker에 대한 처리를 위한 UIImagePickerControllerDelegate, UINavigationControllerDelegate 프로토콜 채택
이미지 피커의 처리를 위한 메서드 imagePickerController(_:info:) 메서드 선언

Hansolkkim committed:
Fix: PhotoView가 터치되면 아무런 반응이 발생하지 않던 이슈 해결

Photo 객체가 Rectangle을 상속받도록 하여, Photo 객체가 됐을 때도 planeDidSpecifiedRectangle(_:) 메서드가 동작하도록 수정

Hansolkkim committed:
Feat: Photo가 Specified됐을 때 배경색 버튼을 랜덤으로 바꾸지 못 하도록 배경색 버튼을 비활성화하는 코드 추가

Hansolkkim committed:
Refactor: BackgroundColor 타입의 타입메서드 random() 수정

r,g,b프로퍼티를 UInt8로 설정했기 때문에, 외부에서 BackgroundColor 타입을 초기화할 때 범위를 알 수 있게 된다 판단해 타입 프로퍼티 possibleColorValues 제거
해당 프로퍼티를 제거했기 때문에 random() 타입 메서드에서 랜덤 value를 생성하는 방법 수정

Hansolkkim committed:
Refactor: 다형성을 이용해 배경색 버튼의 Hidden 여부를 판단하도록 수정

Rectangularable 프로토콜에 backgroundColorButtonShoudBecomeHidden() 메서드 추가
Rectangle 에서는 해당 메서드가 false를, Photo 에서는 해당 메서드가 true를 반환하도록 하여, ViewController에서 배경색 버튼의 hidden 여부를 결정
해당 메서드를 이용하기 위해 updateBackgroundColorButton() 메서드가 Rectangularable 타입의 매개변수를 받도록 수정

godrm commented on DrawingApp/DrawingApp/Model/Photo.swift:
```diff
@@ -0,0 +1,21 @@
+
+import Foundation
+
+class Photo: Rectangle {
+    private(set) var image: Data
+    
+    init(id: ID, size: Size, point: Point, image: Data, alpha: Alpha) {
+        self.image = image
+        super.init(id: id, size: size, point: point, backgroundColor: BackgroundColor.init(r: 0, g: 0, b: 0), alpha: alpha)
```
이렇게 Rectangle을 상속하면 불필요한 backgroundColor 속성도 상속받게 되죠. 
그래서 공통 부분만 상속이 되는 구조가 좋습니다.

Hansolkkim replied:
그럼 먼저 Rectangle에서 채택했던 Rectangularable 프로토콜을 채택하는 더 상위의 클래스를 선언하고, Rectangularable 프로토콜이 가지는 프로퍼티를 Rectangle, Photo가 공통적으로 가지는 부분만 가지도록 줄이고, Rectangle, Photo는 그 상위 클래스를 상속받도록 하는 방법을 사용해보겠습니다.

godrm commented on DrawingApp/DrawingApp/Model/Photo.swift:
```diff
@@ -0,0 +1,21 @@
+    }
+    
+    override func backgroundColorButtonShouldBecomeHidden() -> Bool {
+        return true
```
이런 메소드가 보다는 프로토콜로 추상화를 시도해보세요

Hansolkkim replied:
해당 메서드도 사실 프로토콜로 추상화하려는 목적으로 Rectangularable 프로토콜에서 정의해준 메서드입니다. 그래서 Rectangularable 프로토콜을 채택한 클래스들에서 해당 메서드를 구현하도록 하여 Rectangle, Photo 에서 각각 다른 값을 반환하도록하고, `updateBackgroundColorButton(with:)` 메서드에서 두 객체를 일괄적으로 처리하려고 했습니다.
혹시 이런 프로토콜 추상화 보다는, 각각의 객체가 따로 채택한 프로토콜에서 정의해주도록 하는걸 말씀하시는 걸까요?

Hansolkkim replied:
해당 메서드도 사실 프로토콜로 추상화하려는 목적으로 Rectangularable 프로토콜에서 정의해준 메서드입니다. 그래서 Rectangularable 프로토콜을 채택한 클래스들에서 해당 메서드를 구현하도록 하여 Rectangle, Photo 에서 각각 다른 값을 반환하도록하고, `updateBackgroundColorButton(with:)` 메서드에서 두 객체의 동일한 메서드를 사용해 처리하려고 했습니다.
혹시 이런 프로토콜 추상화 보다는, 각각의 객체가 따로 채택한 프로토콜에서 정의해주도록 하는걸 말씀하시는 걸까요?

godrm commented on DrawingApp/DrawingApp/Model/Plane.swift:
```diff
@@ -8,53 +8,63 @@
-            notificationCenter.post(name: Plane.didSpecifyRectangle, object: self)
-            return .failure(.cannotSpecifyRectangleError)
+        for rectangle in rectangles.reversed() {
+            if rectangle.isPointInArea(point) {
+                self.specifiedRectangle = rectangle
+                NotificationCenter.default.post(name: Plane.NotificationNames.didSpecifyRectangle, object: self, userInfo: [Plane.UserInfoKeys.specifiedRectangle: rectangle])
+                return .success(rectangle)
```
이런 구문도 guard else로 하면 들여쓰지 않고 흐름을 다룰 수도 있습니다. 참고하세요

Hansolkkim replied:
for문 내에 if문을 사용하는 구문 내에 return 및 post 하는 부분이 있어 들여쓰기가 깊어져 읽기가 힘들어졌다는 뜻으로 이해하고,
 해당 부분을 새로운 private 메서드로 만들어서 리턴값을 받아오는 과정에 guard let 구문을 사용하도록 수정해보겠습니다. 
```swift
// class Plane
private func findRectangle(above point: Point) -> AnyRectangularable? {
    for rectangle in rectanlges.reversed() {
      if rectangle.isPointinArea(point) {
        return rectangle
      }
    }
    return nil
}

public func specifyRectangle(point: Point) -> Result<AnyRectangularable, PlaneError> {
    guard let specifiedRectangle = findRectangle(above: point)  else {
        self.specifiedRectangle = nil
        NotificationCenter.default.post(name: Plane.NotificationNames.didSpecifyRectangle, object: self)
        return .failure(.cannotSpecifyRectangleError)
    }
    self.specifiedRectangle = specifiedRectangle
    NotificationCenter.default.post(name: Plane.NotificationNames.didSpecifyRectangle, object: self, userInfo: [Plane.UserInfoKeys.specifiedRectangle: specifiedRectangle])
    return .success(specifiedRectangle)
}
```
위 코드처럼, 들여쓰지 않고 흐름을 다루도록 수정해보겠습니다!

godrm commented on DrawingApp/DrawingApp/Model/Rectangle.swift:
```diff
@@ -8,9 +8,13 @@
 import Foundation
 
 protocol Rectangularable {
+    var size: Size {get}
+    var backgroundColor: BackgroundColor {get}
+    var alpha: Alpha {get}
     func isPointInArea(_ point: Point) -> Bool
     func changeBackgroundColor(to newColor: BackgroundColor)
     func changeAlphaValue(to newAlpha: Alpha)
+    func backgroundColorButtonShouldBecomeHidden() -> Bool
```
이렇게 프로토콜을 한꺼번 선언하면 super class에 다 있는 것과 별 차이가 없습니다
ISP 원칙에 따라서 역할별로 분리해보세요. 그러면 필요한 것과 불필요한 것을 선택할 수 있을 겁니다

Hansolkkim replied:
size, point, alpha 프로퍼티와 isPointInArea(_:), changeAlphaValue(to:) 메서드는 모두 공통적으로 가지는 프로퍼티,메서드 이므로 이 내용만 해당 프로토콜에 남겨놓고, BackgroundColor와 관련된 프로퍼티와 메서드는 다른 프로토콜로 분리해보도록 하겠습니다!

godrm commented on DrawingApp/DrawingApp/Model/ViewFactory.swift:
```diff
@@ -9,7 +9,7 @@ import Foundation
 
 class ViewFactory {
-    static func makeRectangleView(of rectangle: Rectangle) -> RectangleViewable {
+    static func makeRectangleView(of rectangle: Rectangle) -> RectangleViewable & UIView {
```
`RectangleViewable & UIView` 형태는 좋은 접근이네요 
(사소한 거지만 구체 타입은 하나만 나올 수 있으니 앞쪽에 써줘도 좋을 것 같네요)

Hansolkkim replied:
> `RectangleViewable & UIView` 형태는 좋은 접근이네요
> (사소한 거지만 구체 타입은 하나만 나올 수 있으니 앞쪽에 써줘도 좋을 것 같네요)

다른 분들 리뷰를 보다보니, 팩토리 메서드의 반환값도 추상 타입으로 변경해보라는 말씀을 봤었는데, 혹시 UIView와 같이 조금은 추상적인 타입을 반환해도 일반적인 구체타입(RectangleView)을 반환하는 것보다는 더 괜찮은 접근일까요?

godrm replied:
여기서는 아마 UIView가 없으면 addSubview()가 안될꺼에요. 뷰에 대한 super-class 니까 어쩔 수 없는 부분이죠

godrm commented on DrawingApp/DrawingApp/Model/ViewFactory.swift:
```diff
@@ -21,6 +21,19 @@ class ViewFactory {
+    static func makePhotoView(of photo: Photo) -> RectangleViewable & UIView {
+        let point = photo.point
+        let size = photo.size
+        let frame = CGRect(x: point.x, y: point.y, width: size.width, height: size.height)
```
이렇게 할 경우 photo.size가 엄청 크거나 작으면 문제가 되지 않을까요?

Hansolkkim replied:
문제가 될거라고 생각해서 테스트 해봤는데, 문제가 발생하지 않고 자동으로 이미지 크기가 photo.size에 맞게 자동으로 수정이 되네요.. 그래도 오류 발생하지 않도록, 위 메서드 내에 아래 코드를 추가하겠습니다.
```swfit
newView.clipsToBounds = true
newView.contentMode = .scaleAspectFill
```

godrm commented on DrawingApp/DrawingApp/ViewController.swift:
```diff
@@ -110,67 +140,76 @@ class ViewController: UIViewController {
     @objc func planeDidAddRectangle(_ notification: Notification) {
-        guard let addedRectangle = notification.userInfo?[Plane.addedRectangle] as? Rectangle else {return}
+        guard let addedRectangle = notification.userInfo?[Plane.UserInfoKeys.addedRectangle] as? Rectangle else {return}
         os_log("\(addedRectangle)")
         
-        guard let newRectangleView = ViewFactory.makeRectangleView(of: addedRectangle) as? RectangleView else {return}
+        let newRectangleView = ViewFactory.makeRectangleView(of: addedRectangle)
```
사전에 Key값으로 넣으려면 `as? Rectangle` 이 부분이 어쩔 수 없겠네요. 
그런 맥락에서는 구체 타입이 정해져야 하고, 의존적일 수 밖에 없네요
그러면 매핑하는 구조를 바꿀 것인가 또는 전달하는 값을 바꿀 것인가 선택해야겠네요. 
저도 다른 방법을 생각해보고 의견을 남기겠습니다.

godrm replied:
제가 테스트해보니 AnyHashable 자체로 바꾸면 들어가긴 하네요. 
```
        let userInfo : [AnyHashable:Any] = ["A":ObjectA()]
        let object = userInfo["A"] as! AnyHashable
        var map = [AnyHashable: Int]()
        map[object] = 1
```

Hansolkkim replied:
AnyHashable로  타입 캐스팅하는 방법은 생각 못 했네요.. 답변 감사합니다!

godrm reviewed (changes requested):
전체적인 구조로 봐서는 좋습니다. 
프로토콜로 추상화하는 부분도 잘 하셨는데, 프로토콜이 커지고 있어서 그것만 분리해보면 더 좋을 것 같네요. 
다음 내용을 확인해보시고 프로토콜을 ISP 적용해보고 넘어가죠!

Hansolkkim committed:
Refactor: Rectangle, Photo 타입이 공통적으로 가지는 부분만 가지는 부모 클래스 AnyRectangularable 선언

Photo, Rectangle이 AnyRectangularable을 상속받음으로써 상위 모듈인 Plane, RectangleFactory에서는 조금 더 추상적인 타입인 AnyRectangularable을 사용할 수 있도록 수정
Rectangle 클래스에서만 가지는 프로퍼티, 메서드를 추상화한 프로토콜 BackgroundColorChangable 선언하여 Rectangle이 채택하도록 수정
Photo 클래스에서만 가지는 프로퍼티를 추상화한 프로토콜 ImageDataHavable 선언하여 Photo가 채택하도록 수정
AnyRectangularable 타입이 기존의 Rectangularable 프로토콜 타입을 대체한다고 생각하여, 기존에 Rectangularable로 반환하거나, 매개변수로 받는 것들의 타입을 모두 AnyRectangularable 타입으로 수정

Hansolkkim committed:
Refactor: Plane 클래스에 새로운 메서드 findRectangle(above:) 선언

해당 메서드를 사용하여 들여쓰기 깊이를 줄여주어, 흐름을 볼 때 조금 더 읽기 편하도록 수정

Hansolkkim committed:
Fix: 무한루프 발생 이슈에 대한 해결

AnyRectangularable 클래스의 hash 메서드에서 self를 combine해서 무한루프가 발생하는 문제를 해결하기 위해 self.id를 combine하도록 수정

Hansolkkim committed:
Refactor: photo.size가 너무 크거나 작을 경우 발생할 수 있는 PhotoView 생성에서의 문제를 해결하기 위한 코드 추가

Hansolkkim commented:
@godrm 말씀하신 부분 수정해보았습니다! 이번 수정 진행하면서 추상 타입을 이용하고, 추상 타입의 채택한 구체 타입들의 다형성을 이용해 무언가를 처리하도록 하는 방법에 대해서 많이 생각하게 됐던 것 같습니다.
이번 수정 진행하면서 생겼던 질문이 있는데 아래에 작성하겠습니다!
1. ISP 원칙을 지키기 위해, 하나의 객체에서 여러 프로토콜을 채택하는 경우가 많아졌는데, 이런 경우에 특정 behavior를 하는 프로토콜로 객체를 타입 캐스팅해서 사용하고, 또 다른 behavior를 할 때는 해당 프로토콜로 객체를 타입캐스팅해서 사용하는 지점이 많이지게 되었습니다. 
  이렇게 사용하려는 behavior에 맞게 매번 타입 캐스팅을 하는게 일반적인 걸까요?
2. 위에서 Photo.swift에 대한 코드 리뷰를 보다가, 저는 `동일한 프로토콜을 채택한 객체는 동일한 메서드를 자기 나름대로의 행위를 한다

Hansolkkim commented:
@godrm 말씀하신 부분 수정해보았습니다! 이번 수정 진행하면서 추상 타입을 이용하고, 추상 타입의 채택한 구체 타입들의 다형성을 이용해 무언가를 처리하도록 하는 방법에 대해서 많이 생각하게 됐던 것 같습니다.
이번 수정 진행하면서 생겼던 질문이 있는데 아래에 작성하겠습니다!
1. ISP 원칙을 지키기 위해, 하나의 객체에서 여러 프로토콜을 채택하는 경우가 많아졌는데, 이런 경우에 특정 behavior를 하는 프로토콜로 객체를 타입 캐스팅해서 사용하고, 또 다른 behavior를 할 때는 해당 프로토콜로 객체를 타입캐스팅해서 사용하는 지점이 많이지게 되었습니다. 
  이렇게 사용하려는 behavior에 맞게 매번 타입 캐스팅을 하는게 일반적인 걸까요?
2. 위에서 Photo.swift에 대한 코드 리뷰를 보다가, 저는 `동일한 프로토콜을 채택한 객체는 동일한 메서드를 객체마다 다르게 동작한다` 는 관점에서  `backgroundButtonShouldBecomeHidden()` 메서드를 Rectangularable이라는 프로토콜에 정의해주고, 해당 프로토콜을 Rectangle, Photo 가 채택하여 해당 메서드를 다르게 구현하도록 했습니다. 
  그런데 해당 부분에서 프로토콜로 추상화해보라는 말씀을 하셔서, 혹시 어떤 방식의 추상화를 말씀하신건지 알 수 있을까요?

godrm commented:
> @godrm 말씀하신 부분 수정해보았습니다! 이번 수정 진행하면서 추상 타입을 이용하고, 추상 타입의 채택한 구체 타입들의 다형성을 이용해 무언가를 처리하도록 하는 방법에 대해서 많이 생각하게 됐던 것 같습니다. 이번 수정 진행하면서 생겼던 질문이 있는데 아래에 작성하겠습니다!
> 
> 1. ISP 원칙을 지키기 위해, 하나의 객체에서 여러 프로토콜을 채택하는 경우가 많아졌는데, 이런 경우에 특정 behavior를 하는 프로토콜로 객체를 타입 캐스팅해서 사용하고, 또 다른 behavior를 할 때는 해당 프로토콜로 객체를 타입캐스팅해서 사용하는 지점이 많이지게 되었습니다.
>    이렇게 사용하려는 behavior에 맞게 매번 타입 캐스팅을 하는게 일반적인 걸까요?
> 2. 위에서 Photo.swift에 대한 코드 리뷰를 보다가, 저는 `동일한 프로토콜을 채택한 객체는 동일한 메서드를 객체마다 다르게 동작한다` 는 관점에서  `backgroundButtonShouldBecomeHidden()` 메서드를 Rectangularable이라는 프로토콜에 정의해주고, 해당 프로토콜을 Rectangle, Photo 가 채택하여 해당 메서드를 다르게 구현하도록 했습니다.
>    그런데 해당 부분에서 프로토콜로 추상화해보라는 말씀을 하셔서, 혹시 어떤 방식의 추상화를 말씀하신건지 알 수 있을까요?

위 내용을 특정 커밋이나 코드에 코멘트해주면 더 정확하게 짚어볼 수 있을 것 같아요. 바뀐 부분이 많은데 찾기 어려울 수도 있으니까요 

- 일반적으로 답변해보자면 타입 캐스팅을 하지않고 사용하는 인터페이스만 갖는 프로토콜을 고르는 게 가장 좋습니다. 그럴려고 나누는 겁니다

- backgroundButtonShouldBecomeHidden() 부분은 찾아볼게요

godrm commented on DrawingApp/DrawingApp/ViewController.swift:
```diff
@@ -197,19 +197,20 @@ extension ViewController {
     
     @objc func planeDidChangeRectangleBackgroundColor(_ notification: Notification) {
-        guard let backgroundColorChangedRectangle = notification.userInfo?[Plane.UserInfoKeys.changedRectangle] as? Rectangle,
+        guard let backgroundColorChangedRectangle = notification.userInfo?[Plane.UserInfoKeys.changedRectangle] as? AnyRectangularable,
```
Notification에서 userInfo로 전달되는 부분은 타입 정보가 사라지니까 타입 캐스팅이 필요하겠네요. 
이런 부분이 NotificationCenter를 쓰면 불편한 부분이죠.

godrm commented on DrawingApp/DrawingApp/Model/Photo.swift:
```diff
@@ -7,15 +7,19 @@
-    init(id: ID, size: Size, point: Point, image: Data, alpha: Alpha) {
-        self.image = image
-        super.init(id: id, size: size, point: point, backgroundColor: BackgroundColor.init(r: 0, g: 0, b: 0), alpha: alpha)
+    override var backgroundColorButtonShouldBecomeHidden: Bool {
+        return true
     }
```
backgroundColorButtonShouldBecomeHidden 이군요. 본문에서 언급한 것도 본인도 다른걸로 쓰셨네요 ㅎㅎ (Color가 빠짐)
이래서 코드에 코멘트하는 게 정확합니다 

---
맞아요. 이것도 다형성이죠. 그렇지만 이렇게 새로운 것을 추가하기 보다는 특정한 Behavior를 가진 인터페이스가 있느냐 없느냐로도 검증할 수 있습니다. 저렇게 되면 타입이 추가될 때마다 저런 메소드를 다른 타입에도 추가해줘야 하니까요. 
Image가 추가될 때 기존 Rectangle은 영향을 안 받는 게 좋다는 겁니다.

Hansolkkim replied:
아.. 이런 점에서 추상화를 말씀하신거군요.. 감사합니다!

Hansolkkim commented on DrawingApp/DrawingApp/ViewController.swift:
```diff
@@ -197,19 +197,20 @@ extension ViewController {
-        
-        let newBackgroundColor = backgroundColorChangedRectangle.backgroundColor
-        matchedView.changeBackgroundColor(to: newBackgroundColor.convertToUIColor())
+        if let backgroundColorChangedRectangle = backgroundColorChangedRectangle as? BackgroundColorChangable {
+            let newBackgroundColor = backgroundColorChangedRectangle.backgroundColor
+            matchedView.changeBackgroundColor(to: newBackgroundColor.convertToUIColor())
+        }
```
질문 1번에 해당하는 부분입니다. 
userInfo로 가져온 내용이 Photo냐 Rectangle이냐에 따라 ImageDataHavable프로토콜을 채택했냐 BackgroundColorChangable 프로토콜을 채택했냐로 바뀌게 되고, 각각 프로토콜이 가지고 있는 프로퍼티, 메서드가 달라 이런 방식으로 하나하나 타입캐스팅을 사용해주도록 했는데, 
ISP를 지키도록 하면 프로토콜을 각각의 역할대로 나누어야 할 것 같고, 그렇게되면 같은 프로토콜을 채택했지만 다른 프로토콜도 채택하여 각각 분기가 필요하게 되는 부분이 있어, 분기 없이 이것을 처리하려면 어떻게 판단하여 처리하는게 적절할지 헷갈리네요..ㅠㅠ

godrm replied:
프로토콜을 역할별로 나누면 A가 P1, P2, P3 프로토콜을 채택한 상태니까, 그 중에서 필요한 프로토콜들을 조합해서 접근하면 되는 거죠 
적합한 프로토콜만 선택하면 분기가 필요하지 않아야 합니다. 특히 여기 코드에서는 드러나지 않아서 모르겠습니다.

Hansolkkim replied:
답변과 리뷰 감사합니다! 다음 스텝 진행할 때 해당 부분 조금 더 고민 후 수정해보도록 하겠습니다!
